prow/sidecar/run.go

```diff
@@ -150,29 +150,34 @@ func (o Options) Run(ctx context.Context, logFile *os.File) (int, error) {
 
 	o.preUpload()
 
-	buildLogs := logReaders(entries)
+	buildLogs := logReadersFuncs(entries)
 	metadata := combineMetadata(entries)
 	return failures, o.doUpload(context.Background(), spec, passed, aborted, metadata, buildLogs, logFile, &once)
 }
 
 const errorKey = "sidecar-errors"
 
-func logReaders(entries []wrapper.Options) map[string]io.Reader {
-	readers := make(map[string]io.Reader)
+func logReadersFuncs(entries []wrapper.Options) map[string]gcs.ReaderFunc {
+	readerFuncs := make(map[string]gcs.ReaderFunc)
 	for _, opt := range entries {
+		opt := opt
+		f := func() (io.ReadCloser, error) {
+			log, err := os.Open(opt.ProcessLog)
+			if err != nil {
+				logrus.WithError(err).Errorf("Failed to open %s", opt.ProcessLog)
```

Comment:
I see you carried this over, but this seems weird to me, shouldn't we just return the error so the caller can retry?

Reply:
It sounds reasonable but we are going to change the behavior here. If we do that then the code will keep retrying if it fails to open the file, instead of uploading the error "Failed to open {FILE}: {ERROR}" as it is now. I am ok with that though.

Reply:
Yeah good point. Lets maybe just keep it as-is for now to fix the retrying and have a follow-up discussion about changing behavior here